passes/techmap/clockgate.cc

```diff
@@ -37,6 +39,161 @@ ClockGateCell icg_from_arg(std::string& name, std::string& str) {
 	return c;
 }
 
+static std::pair<std::optional<ClockGateCell>, std::optional<ClockGateCell>>
+	find_icgs(std::string filename) {
+	std::ifstream f;
+	f.open(filename.c_str());
+	if (f.fail())
+		log_cmd_error("Can't open liberty file `%s': %s\n", filename.c_str(), strerror(errno));
+	LibertyParser libparser(f);
+	f.close();
+	auto ast = libparser.ast;
+
+	// We will pick the most suitable ICG absed on tie_lo count and area
+	struct ICGRankable {
+		ClockGateCell icg;
+		double area;
+	};
+	std::optional<ICGRankable> best_pos;
+	std::optional<ICGRankable> best_neg;
+
+	if (ast->id != "library")
+		log_error("Format error in liberty file.\n");
+
+	// This is a lot of boilerplate, isn't it?
+	for (auto cell : ast->children)
+	{
+		if (cell->id != "cell" || cell->args.size() != 1)
+			continue;
+
+		const LibertyAst *dn = cell->find("dont_use");
+		if (dn != nullptr && dn->value == "true")
+			continue;
+
+		// TODO Should we have user-specified dont_use_cells here?
+
+		const LibertyAst *icg_kind_ast = cell->find("clock_gating_integrated_cell");
+		if (icg_kind_ast == nullptr)
+			continue;
+
+		auto cell_name = cell->args[0];
+		auto icg_kind = icg_kind_ast->value;
+		std::vector<std::string> kind_tags;
+		std::stringstream ss(icg_kind);
+		std::string tag;
+		while (std::getline(ss, tag, '_')) {
+			kind_tags.push_back(tag);
+		}
+		if ((kind_tags.size() < 2) || (kind_tags.size() > 4))
+			log_error("Malformed liberty file - invalid clock_gating_integrated_cell value %s in cell %s\n",
```

Comment:
Are the set of valid `clock_gating_integrated_cell` standardized? I am a bit worried by the different "malformed input" errors we throw here. I am not sure Liberty can't get extended over time, even if the current standard restricts the valid values to a closed set

Reply:
The spec says "A concatenation of up to four strings that describe the functionality of the
cell to the clock-gating software". If it's less than two, we're missing the none vs latch etc detail, or whether it's posedge or negedge. We can't go on without those and in fact it means the Liberty file has non-sensical cells in it. I don't think we should just ignore non-sensical cells